client/src/components/Upload/UploadModal.vue

```diff
@@ -259,89 +82,9 @@ const UploadModal = {
-                files: [],
-                error_message: null,
-            };
-            // add upload tools input data
-            if (items && items.length > 0) {
-                var inputs = {
-                    file_count: items.length,
-                    dbkey: items[0].get("genome", "?"),
-                    // sometimes extension set to "" in automated testing after first upload of
-                    // a session. https://github.com/galaxyproject/galaxy/issues/5169
-                    file_type: items[0].get("extension") || "auto",
-                };
-                for (var index in items) {
-                    var it = items[index];
-                    it.set("status", "running");
-                    if (it.get("file_size") > 0) {
-                        var prefix = `files_${index}|`;
-                        inputs[`${prefix}type`] = "upload_dataset";
-                        if (it.get("file_name") != "New File") {
-                            inputs[`${prefix}NAME`] = it.get("file_name");
-                        }
-                        inputs[`${prefix}space_to_tab`] = (it.get("space_to_tab") && "Yes") || null;
-                        inputs[`${prefix}to_posix_lines`] = (it.get("to_posix_lines") && "Yes") || null;
-                        inputs[`${prefix}dbkey`] = it.get("genome", null);
-                        inputs[`${prefix}file_type`] = it.get("extension", null);
-                        let uri;
-                        let how;
-                        switch (it.get("file_mode")) {
-                            case "new":
-                                inputs[`${prefix}url_paste`] = it.get("url_paste");
-                                break;
-                            case "ftp":
-                                uri = it.get("file_path");
-                                how = "ftp_files";
-                                if (uri.indexOf("://") >= 0) {
-                                    how = "url_paste";
-                                }
-                                inputs[`${prefix}${how}`] = uri;
-                                break;
-                            case "local":
-                                data.files.push({
-                                    name: `${prefix}file_data`,
-                                    file: it.get("file_data"),
-                                });
-                        }
-                    } else if (it.get("optional")) {
-                        continue;
-                    } else {
-                        data.error_message = "Upload content incomplete.";
-                        it.set("status", "error");
-                        it.set("info", data.error_message);
-                        break;
-                    }
-                }
-                data.payload.inputs = JSON.stringify(inputs);
-            }
-            return data;
-        },
+    },
+    created() {
+        this.eventHub.$on("upload:show", this.show);
```

Comment:
We do create multiple upload modals with different props and callbacks, one is in [DataDialog.vue](https://github.com/Nerdinacan/galaxy/blob/1ebdb8fb00f47e7a754322b315558614d67b4c49/client/src/components/DataDialog/DataDialog.vue#L138), and they all get triggered with `upload:show`. If you want to try this, open a tool, click on the button next to a data input, and at the bottom you have the option to upload a dataset.
This doesn't currently show anything since https://github.com/Nerdinacan/galaxy/blob/1ebdb8fb00f47e7a754322b315558614d67b4c49/client/src/components/DataDialog/DataDialog.vue#L136 passes in `modalShow: true` which is not a prop anymore.

I don't know what the solution is there, a single global upload button may work, but then we need to manage props. Maybe custom upload modals are a better solution ?

Reply:
Hmmm let me look carefully at that. There was only one "instance" of an upload modal as I understood it. It was simply being shown and hidden in a byzantine and unexplained train of props. But, you're telling me the actual goal was to have multiple upload modals open at once?

Reply:
goal is a big word, but the callback and the `modalShow` prop made it possible to upload datasets straight from an open tool form or workflow run form and have the form update its selected dataset. If we can adapt the current singleton or a variant of the upload modal for those purposes that might be fine too.

Reply:
Got it, I'll try to create a good test scenario and a fix while still letting us initialize the Upload in a sane way.

Reply:
Alrighty, I found the PR that John originally opened for that functionality.
https://github.com/galaxyproject/galaxy/pull/9809

He's got a nice little video of his goal which I think is an awesome way to describe this stuff. But the way to implement this kind of behavior is with a promise. You don't want to assume everything you need is necessarily there. That's why feature is broken in dev, and that's why the selenium tests are failing (because the inits are not completed yet.)

Instead of this:

```js
 /** Called when user decides to upload new data */
  onUpload() {
      const propsData = {
          multiple: this.multiple,
          format: this.format,
          callback: this.callback,
          modalShow: true,
      };
      mountUploadModal(propsData);
      this.modalShow = false;
  },
```

It should really be something more like:
```js
onUpload() {
  const propsData = {
      multiple: this.multiple,
      format: this.format,
      modalShow: true,
  };
  mountUploadModal(propsData).then(() => {
      this.callback();
  })
}
```

Reply:
I'll rewrite it so that the mount function can return a promise when needed.

Reply:
Avoiding the Function prop value was turning into a rabbit-hole. I have a complete jquery-free, select2-free, Galaxy global var-free replacement for Upload almost ready, so I just restored what was there for now.